### PR TITLE
Add the Units library

### PR DESCRIPTION
**Pull request by rdeterre**
This merge request is to add the Units library
https://github.com/nholthaus/units

Specify library name and version:  **units/2.3.1**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/units/all/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 2.8.11)
+project(unitstest CXX)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(${PROJECT_NAME} ${PROJECT_NAME}.cpp)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
+set_target_properties(${PROJECT_NAME} PROPERTIES CXX_STANDARD 11)
```
Comment (madebr) [inline recipes/units/all/test_package/CMakeLists.txt]: The library requires c++14
```suggestion
set_target_properties(${PROJECT_NAME} PROPERTIES CXX_STANDARD 14)
```

File: recipes/units/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    homepage = "https://github.com/nholthaus/units"
+    description = "A compile-time, header-only, dimensional analysis and unit conversion library built on c++14 with no dependencies"
+    topics = ("unit-conversion", "dimensional-analysis", "cpp14",
+              "template-metaprogramming", "compile-time", "header-only",
+              "no-dependencies")
+    license = "MIT"
+    no_copy_source = True
+
+    _source_subfolder = "source_subfolder"
+
```
Comment (madebr) [inline recipes/units/all/conanfile.py]: May be add a `configure()` method here that checks the current c++ standard.
If `self.settings.compiler.cppstd` is defined, check that it is >= 14.
Also check whether `self.settings.compiler.version` >= minimum version of that compiler supporting c++14

File: recipes/units/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def package(self):
+        self.copy("units.h",
+                  dst="include/units",
```
Comment (madebr) [inline recipes/units/all/conanfile.py]: The cmake script does not copy the headers in a `units` subdirectory.
https://github.com/nholthaus/units/blob/9f09abb32641fdf8dbb2c49ee81717d434624218/CMakeLists.txt#L230-L232
```suggestion
                  dst="include",
```

File: recipes/units/all/test_package/CMakeLists.txt (hunk 1 above)
Comment (rdeterre) [reply recipes/units/all/test_package/CMakeLists.txt]: Thanks, I applied this change

File: recipes/units/all/conanfile.py (hunk 2 above)
Comment (rdeterre) [reply recipes/units/all/conanfile.py]: Thanks. I wasn't exactly sure how to check by compiler which one supports C++14, so instead I check for compiler versions as specified on the unit library's README file.

File: recipes/units/all/conanfile.py (hunk 3 above)
Comment (rdeterre) [reply recipes/units/all/conanfile.py]: Thanks. I took the subdirectory out.

File: recipes/units/all/test_package/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    generators = "cmake"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        os.chdir("bin")
+        self.run(".%sunitstest" % os.sep)
```
Comment (madebr) [inline recipes/units/all/test_package/conanfile.py]:
```suggestion
        if not tools.cross_building(self.settings):
            bin_path = os.path.join("bin", "unitstest")
            self.run(bin_path, run_environment=True)
```
Support cross building

File: recipes/units/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    _source_subfolder = "source_subfolder"
+
+    def _compiler_is_supported(self):
+        supported_compilers = [("clang", 3.4), ("gcc", "4.9.3"),
+                               ("Visual Studio", 14.0)]
+        compiler, version = self.settings.compiler, Version(
+            self.settings.compiler.version)
+        return any(compiler == sc[0] and version >= sc[1]
+                   for sc in supported_compilers)
```
Comment (madebr) [inline recipes/units/all/conanfile.py]:
```suggestion
        if self.settings.compiler.cppstd:
            tools.check_min_cppstd(self, "14")
        minimum_version = {
            "clang": 3.4,
            "gcc": "4.9.3",
            "Visual Studio", 14.0,
        }.get(str(self.settings.compiler))
        if not minimum_version:
             self.output.warn("Unknown compiler: assuminging compiler supports c++14")
        else:
             if Version(self.settings.compiler) < minimum_version:
                 raise ConanInvalidConfiguration("Compiler does not support c++14")
```
I think this code is a bit more readable.
Also, untested.

File: recipes/units/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    def package(self):
+        self.copy(
+            "units.h",
+            dst="include",
+            src=os.path.join(self._source_subfolder, "include"))
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+
+    def package_info(self):
+        self.info.header_only()
```
Comment (madebr) [inline recipes/units/all/conanfile.py]:
```suggestion
    def package_id(self):
        self.info.header_only()
```

File: recipes/units/all/test_package/conanfile.py (hunk 4 above)
Comment (rdeterre) [reply recipes/units/all/test_package/conanfile.py]: Thanks, this change is in.

File: recipes/units/all/conanfile.py (hunk 5 above)
Comment (rdeterre) [reply recipes/units/all/conanfile.py]: Got it, I put this in.

File: recipes/units/all/conanfile.py (hunk 6 above)
Comment (rdeterre) [reply recipes/units/all/conanfile.py]: Thanks, this is in.

**Comment by uilianries** [comment]
There is a massive number of public definitions which could be imported to this recipe: https://github.com/nholthaus/units/blob/master/CMakeLists.txt#L89

e.g.

```python
options = {"predefined_units": [True, False]}
default_options = {"predefined_units": True}

....

def package_info(self):
    if not self.predefined_units:
        self.cpp_info.defines.append("DISABLE_PREDEFINED_UNITS")
```

**Comment by madebr** [comment]
> There is a massive number of public definitions which could be imported to this recipe: https://github.com/nholthaus/units/blob/master/CMakeLists.txt#L89

Nothing what a regex can't solve. ™️  🤣 

```python
set(l.lower() for l in re.findall(r"if\(ENABLE_([A-Z_]+)\)", open(os.path.join(self._source_subfolder, "CMakeLists.txt")).read()))
```
==>
```python
{'predefined_illuminance_units', 'predefined_volume_units', 'predefined_substance_units', 'predefined_torque_units', 'predefined_energy_units', 'predefined_angle_units', 'predefined_radiation_units', 'predefined_magnetic_field_strength_units', 'predefined_voltage_units', 'predefined_current_units', 'predefined_acceleration_units', 'predefined_area_units', 'predefined_solid_angle_units', 'predefined_pressure_units', 'predefined_luminous_flux_units', 'predefined_data_transfer_rate_units', 'predefined_density_units', 'predefined_inductance_units', 'predefined_charge_units', 'predefined_data_units', 'predefined_temperature_units', 'predefined_capacitance_units', 'predefined_length_units', 'predefined_concentration_units', 'predefined_power_units', 'predefined_velocity_units', 'predefined_force_units', 'predefined_magnetic_flux_units', 'predefined_impedance_units', 'predefined_time_units', 'predefined_luminous_intensity_units', 'predefined_frequency_units', 'predefined_angular_velocity_units', 'predefined_mass_units'}
```

**Comment by uilianries** [comment]
> @uilianries
> 
> > There is a massive number of public definitions which could be imported to this recipe: [nholthaus/units:CMakeLists.txt@`master`#L89](https://github.com/nholthaus/units/blob/master/CMakeLists.txt#L89) [](/nholthaus/units/blob/HEAD@{2020-03-05T18:36:45Z}/CMakeLists.txt#L89)
> 
> Nothing what a regex can't solve. tm rofl
> 
> ```python
> set(l.lower() for l in re.findall(r"if\(ENABLE_([A-Z_]+)\)", open(os.path.join(self._source_subfolder, "CMakeLists.txt")).read()))
> ```
> 
> ==>
> 
> ```python
> {'predefined_illuminance_units', 'predefined_volume_units', 'predefined_substance_units', 'predefined_torque_units', 'predefined_energy_units', 'predefined_angle_units', 'predefined_radiation_units', 'predefined_magnetic_field_strength_units', 'predefined_voltage_units', 'predefined_current_units', 'predefined_acceleration_units', 'predefined_area_units', 'predefined_solid_angle_units', 'predefined_pressure_units', 'predefined_luminous_flux_units', 'predefined_data_transfer_rate_units', 'predefined_density_units', 'predefined_inductance_units', 'predefined_charge_units', 'predefined_data_units', 'predefined_temperature_units', 'predefined_capacitance_units', 'predefined_length_units', 'predefined_concentration_units', 'predefined_power_units', 'predefined_velocity_units', 'predefined_force_units', 'predefined_magnetic_flux_units', 'predefined_impedance_units', 'predefined_time_units', 'predefined_luminous_intensity_units', 'predefined_frequency_units', 'predefined_angular_velocity_units', 'predefined_mass_units'}
> ```

Regex is life! :joy: :joy:

**Pull request merged**